stan/math/rev/core/vari.hpp

```diff
@@ -179,11 +165,375 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
+   * Return the number of rows for this class's `val_` member
+   */
+  const Eigen::Index rows() const { return val_.rows(); }
+  /**
+   * Return the number of columns for this class's `val_` member
+   */
+  const Eigen::Index cols() const { return val_.rows(); }
+  /**
+   * Return the size of this class's `val_` member
+   */
+  const Eigen::Index size() const { return val_.size(); }
+
+  virtual void chain() {}
+  /**
+   * Initialize the adjoint for this (dependent) variable to 1.
+   * This operation is applied to the dependent variable before
+   * propagating derivatives, setting the derivative of the
+   * result with respect to itself to be 1.
+   */
+  inline void init_dependent() { adj_.setOnes(); }
+
+  /**
+   * Set the adjoint value of this variable to 0.  This is used to
+   * reset adjoints before propagating derivatives again (for
+   * example in a Jacobian calculation).
+   */
+  inline void set_zero_adjoint() final { adj_.setZero(); }
+
+  /**
+   * Insertion operator for vari. Prints the current value and
+   * the adjoint value.
+   *
+   * @param os [in, out] ostream to modify
+   * @param v [in] vari object to print.
+   *
+   * @return The modified ostream.
+   */
+  friend std::ostream& operator<<(std::ostream& os, const vari_value<T>* v) {
+    return os << "val: \n" << v->val_ << " \nadj: \n" << v->adj_;
+  }
+
+  /**
+   * Allocate memory from the underlying memory pool.  This memory is
+   * is managed as a whole externally.
+   *
+   * Warning: Classes should not be allocated with this operator
+   * if they have non-trivial destructors.
+   *
+   * @param nbytes Number of bytes to allocate.
+   * @return Pointer to allocated bytes.
+   */
+  static inline void* operator new(size_t nbytes) {
+    return ChainableStack::instance_->memalloc_.alloc(nbytes);
+  }
+
+  /**
+   * Delete a pointer from the underlying memory pool.
+   *
+   * This no-op implementation enables a subclass to throw
+   * exceptions in its constructor.  An exception thrown in the
+   * constructor of a subclass will result in an error being
+   * raised, which is in turn caught and calls delete().
+   *
+   * See the discussion of "plugging the memory leak" in:
+   *   http://www.parashift.com/c++-faq/memory-pools.html
+   */
+  static inline void operator delete(void* /* ignore arg */) { /* no op */
+  }
+
+ private:
+  template <typename>
+  friend class var_value;
+  template <typename, typename>
+  friend class vari_value;
+};
+
+/**
+ * The variable implementation for Eigen sparse matrix types.
+ *
+ * This class is complete (not abstract) and may be used for
+ * constants.
+ *
+ * A variable implementation is constructed with a constant
+ * value. It also stores the adjoint for storing the partial
+ * derivative with respect to the root of the derivative tree.
+ *
+ */
+template <typename T>
+class vari_value<T, std::enable_if_t<is_eigen_sparse_base<T>::value>>
+    : public vari_base, chainable_alloc {
+ public:
+  using PlainObject
```

Comment:
Do we need to assert things aren't expressions or does the enable_if handle that?

Reply:
The sparse impl also inherits from `chainable_alloc`

Reply:
The `enable_if` just checks if the type `T` inherits from `Eigen::SparseMatrixBase`. We still want to do an assertion that the template type is a plain eigen object